Docs (GraphQL): 21.03 feature documentation for federation (#104)
dgraph-io/dgraph-docs: content/graphql/federation/index.md

+++
title = "Apollo Federation"

> weight = 14

Reviewer: Please add a "description" under "title" to give us a description for SEO. Suggestion:

```
description = "Dgraph now supports Apollo federation so that you can create a gateway GraphQL service that includes the Dgraph GraphQL API and other GraphQL services."

Reviewer: Thanks for adding this!

[menu.main]
  name = "Apollo Federation"
  identifier = "federation"
  parent = "graphql"
+++

> Starting from release version 21.03 support has been added for Apollo federation. Hence making it possible to use Dgraph GraphQL API with other GraphQL services through a single gateway.

Reviewer: ```suggestion
Dgraph supports Apollo federation starting in release version 21.03. This lets you create a gateway GraphQL service that includes the Dgraph GraphQL API and other GraphQL services.
```

Reviewer: Edited to remove passive voice

## Support for Apollo federation directives

> The current implementation supports 3 directives, namely `@key`, `@extends`, and `@external`.

Reviewer: ```suggestion
The current implementation supports the following three directives: `@key`, `@extends`, and `@external`.
```

### `@key` directive

> This directive takes one field argument inside it which is called `@key` field. There are few limitations on how to use `@key` directives.

Reviewer: ```suggestion
This directive takes one field argument inside it: the `@key` field. There are few limitations on how to use `@key` directives:
```

> User can define `@key` directive only once for a type, Support for multiple key types is not provided yet.

Reviewer: The items on lines 19 and 20 should be three bullets, IMO (slight suggested rewording):

* Users can define the `@key` directive only once for a type
* Support for multiple key types is not currently available.
* Since the `@key` field acts as a foreign key to resolve entities from the service where it is extended, the field provided as an argument inside the `@key` directive should be of `ID` type or have the `@id` directive on it.

## Mutation for `extended` types
If you want to add an object of `Astronaut` type which is extended in this service.
The mutation `addAstronaut` takes `AddAstronautInput` which is generated as -

```graphql
input AddAstronautInput {
	id: ID!
	missions: [MissionRef]
}
```

Even though the `id` field is of `ID` type which should be ideally generated internally by Dgraph. In this case, it should be provided as input since currently federated mutations aren't supported. The user should provide the value of `id` same as the value present in the GraphQL service where the type  `Astronaut` is defined.

For example, let's take that the type `Astronaut` is defined in some other service `AstronautService` as -

```graphql
type Astronaut @key(fields: "id") {
    id: ID! 
    name: String!
}
```

When adding an object of type `Astronaut`, first it should be added into `AstronautService` service and then the `addAstronaut` mutation should be called with value of `id` provided as an argument which must be equal to the value in `AstronautService` service.

> Use the admin endpoint to query for the generated schema -

Reviewer 2: Let's remove the `getGQLSchema` thing from here, it is not relevant.

```graphql
{
  getGQLSchema {
    generatedSchema
  }
}
```

{{% notice "note" %}}

> Currently we only support federated queries, not federated mutations. We might support it in the future.

Reviewer 2: The apollo spec itself doesn't support federated mutations. So, I don't think we should even say that we might support them later.

{{% /notice %}}

> ## Gateway supported directives

Reviewer 2: I feel this section is also irrelevant to users. They shouldn't be concerned about the bugs in the underlying libs. They only want to use the feature.
So, let's remove this section too.

For example -

```graphql
type Review @key(fields: "id") {
  product: Product @provides(fields: "name price")
}

extend type Product @key(fields: "upc") {
  upc: String @external
  name: String @external
  price: Int @external
}
```

While fetching `Review.product` from the `review` service, and if the `name` or `price` is also queried, the gateway will fetch these from the `review` service itself, meaning that it also resolves these fields, even though both fields are `@external`.

### `@requires` directive
This directive is used on a field to annotate the fieldSet of the base type. It is used to develop a query plan where the required fields may not be needed by the client, but the service may need additional information from other services. 

For example -

```graphql
extend type User @key(fields: "id") {
  id: ID! @external
  email: String @external
  reviews: [Review] @requires(fields: "email")
}
```

When the gateway fetches `user.reviews` from the `review` service, the gateway will get `user.email` from the `User` service and provide it as an argument to the `_entities` query.

> 

Reviewer 2: About `@requires`, we also need to add that:
```
If you feel the need to use `@requires`, you probably also want to add custom logic on such fields. You can do that using `@lambda`/`@custom(http: {...})`
```
See, this PR for example usage of `@requires` with `@lambda` and accordingly add an example here with `@lambda`: https://github.com/dgraph-io/dgraph/pull/7565

## Mutation for `extended` types
If you want to add an object of `Astronaut` type which is extended in this service.
The mutation `addAstronaut` takes `AddAstronautInput` which is generated as -

```graphql
input AddAstronautInput {
	id: ID!
	missions: [MissionRef]
}
```

Even though the `id` field is of `ID` type which should be ideally generated internally by Dgraph. In this case, it should be provided as input since currently federated mutations aren't supported. The user should provide the value of `id` same as the value present in the GraphQL service where the type  `Astronaut` is defined.

For example, let's take that the type `Astronaut` is defined in some other service `AstronautService` as -

```graphql
type Astronaut @key(fields: "id") {
    id: ID! 
    name: String!
}
```

When adding an object of type `Astronaut`, first it should be added into `AstronautService` service and then the `addAstronaut` mutation should be called with value of `id` provided as an argument which must be equal to the value in `AstronautService` service.

> Remember to add the `Extended Definitions` block from the generated schema of Dgraph GraphQL to the other GraphQL services schemas that are part of the gateway. This needs to be done otherwise the Apollo gateway will throw an error like `Custom directives must be implemented in every service.`

Reviewer 3: We don't need to append `Extended Definitions` now. Please remove this block.

Author: Done

+++
title = "Apollo Federation"
description = "Dgraph now supports Apollo federation so that you can create a gateway GraphQL service that includes the Dgraph GraphQL API and other GraphQL services."
weight = 14
[menu.main]
  name = "Apollo Federation"
  identifier = "federation"
  parent = "graphql"
+++

Dgraph supports [Apollo federation](https://www.apollographql.com/docs/federation/) starting in release version 21.03. This lets you create a gateway GraphQL service that includes the Dgraph GraphQL API and other GraphQL services.

## Support for Apollo federation directives

The current implementation supports the following five directives: `@key`, `@extends`, `@external`, `@provides`, and `@requires`.

### `@key` directive
This directive takes one field argument inside it: the `@key` field. There are few limitations on how to use `@key` directives:

> - Users can define the @key directive only once for a type

Reviewer: ```suggestion
- Users can define the `@key` directive only once for a type
```

- Support for multiple key fields is not currently available.

> - Since the @key field acts as a foreign key to resolve entities from the service where it is extended, the field provided as an argument inside the @key directive should be of ID type or have the @id directive on it.

Reviewer: ```suggestion
- Since the `@key` field acts as a foreign key to resolve entities from the service where it is extended, the field provided as an argument inside the `@key` directive should be of `ID` type or have the `@id` directive on it.
```

For example -

```graphql
type User @key(fields: "id") {
   id: ID!
  name: String
}
```

### `@extends` directive

> This directive is provided to give support for extended definitions. Suppose the above defined `User` type is defined in some other service. Users can extend it in Dgraph's GraphQL service by using this directive. 

Reviewer: ```suggestion
This directive provides support for extended definitions. For example, if the above-defined `User` type is defined in some other service, you can extend it in Dgraph's GraphQL service by using the `@extends` directive, as follows: 
```

```graphql
type User @key(fields: "id") @extends{
  id: String! @id @external
  products: [Product]
}
```

> The same is also achievable with the `extend` keyword, i.e., user has the choice to choose between `extend type User ...` or `type User @extends ...`.

Reviewer: ```suggestion
You can also achieve this with the `extend` keyword; so you have a choice between two types of syntax to extend a type into your Dgraph GraphQL service: `extend type User ...` or `type User @extends ...`.
```

### `@external` directive

> This directive is used when the given field is not stored in this service. It can only be used on extended type definitions. As it is used above on the `id` field of `User` type.

Reviewer: ```suggestion
You use this directive when the given field is not stored in this service. It can only be used on extended type definitions. For example, it is used in the example shown above on the `id` field of the `User` type.
```

### `@provides` directive

> This directive is used on a field that tells the gateway to return a specific fieldSet from the base type while fetching the field. 

Reviewer: ```suggestion
You use this directive on a field that tells the gateway to return a specific `FieldSet` from the base type while fetching the field. 
```

Reviewer: Per https://www.apollographql.com/docs/federation/federation-spec/#scalar-_fieldset, we should either write "fieldset" or `_FieldSet`. The first one makes more sense to me in this context, but I could be wrong.

For example -

```graphql
type Review @key(fields: "id") {
  product: Product @provides(fields: "name price")
}

extend type Product @key(fields: "upc") {
  upc: String @external
  name: String @external
  price: Int @external
}
```

> While fetching `Review.product` from the `review` service, and if the `name` or `price` is also queried, the gateway will fetch these from the `review` service itself, meaning that it also resolves these fields, even though both fields are `@external`.

Reviewer: ```suggestion
While fetching `Review.product` from the `review` service, and if the `name` or `price` is also queried, the gateway will fetch these from the `review` service itself. So, the `review` service also resolves these fields, even though both fields are `@external`.
```

Reviewer: Breaking this into two sentences makes it easier to parse (and machine-translate)

### `@requires` directive

> This directive is used on a field to annotate the fieldSet of the base type. It is used to develop a query plan where the required fields may not be needed by the client, but the service may need additional information from other services. 

Reviewer: ```suggestion
You use this directive on a field to annotate the fieldset of the base type. You can use it to develop a query plan where the required fields may not be needed by the client, but the service may need additional information from other services. 
```

For example -

```graphql
extend type User @key(fields: "id") {
  id: ID! @external
  email: String @external
  reviews: [Review] @requires(fields: "email")
}
```

When the gateway fetches `user.reviews` from the `review` service, the gateway will get `user.email` from the `User` service and provide it as an argument to the `_entities` query.

> Using `@requires` alone on a field doesn't make much sense. In cases, where you feel the need to use `@requires`, you would want some custom logic on that field too. You can achieve that using `@lambda` or `@custom(http: {...})` directives.

Reviewer: ```suggestion
Using `@requires` alone on a field doesn't make much sense. In cases where you need to use `@requires`, you should also add some custom logic on that field. You can add such logic using the `@lambda` or `@custom(http: {...})` directives.
```

  "User.reviews": userReviews
})
```

## Generated queries and mutations

In this section, you will see what all queries and mutations will be available to individual service and to the Apollo gateway. 

Let's take the below schema as an example -

```graphql
type Mission @key(fields: "id") {
    id: ID!
    crew: [Astronaut]
    designation: String!
    startDate: String
    endDate: String
}

type Astronaut @key(fields: "id") @extends {
    id: ID! @external
    missions: [Mission]
}
```

The queries and mutations which are exposed to the gateway are -

```graphql
type Query {
	getMission(id: ID!): Mission
	queryMission(filter: MissionFilter, order: MissionOrder, first: Int, offset: Int): [Mission]
	aggregateMission(filter: MissionFilter): MissionAggregateResult
}

type Mutation {
	addMission(input: [AddMissionInput!]!): AddMissionPayload
	updateMission(input: UpdateMissionInput!): UpdateMissionPayload
	deleteMission(filter: MissionFilter!): DeleteMissionPayload
	addAstronaut(input: [AddAstronautInput!]!): AddAstronautPayload
	updateAstronaut(input: UpdateAstronautInput!): UpdateAstronautPayload
	deleteAstronaut(filter: AstronautFilter!): DeleteAstronautPayload
}
```

> The queries for `Astronaut` are not exposed to the gateway since it will be resolved through the `_entities` resolver. Although these queries will be available on the Dgraph GraphQL API endpoint.

Reviewer: ```suggestion
The queries for `Astronaut` are not exposed to the gateway because they are resolved through the `_entities` resolver. However, these queries are available on the Dgraph GraphQL API endpoint.
```

## Mutation for `extended` types
If you want to add an object of `Astronaut` type which is extended in this service.

> The mutation `addAstronaut` takes `AddAstronautInput` which is generated as -

Reviewer: ```suggestion
The mutation `addAstronaut` takes `AddAstronautInput`, which is generated as follows:
```

```graphql
input AddAstronautInput {
	id: ID!
	missions: [MissionRef]
}
```

> Even though the `id` field is of `ID` type which should be ideally generated internally by Dgraph. In this case, it should be provided as input since currently federated mutations aren't supported. The user should provide the value of `id` same as the value present in the GraphQL service where the type  `Astronaut` is defined.

Reviewer: ```suggestion
The `id` field is of `ID` type, which is usually generated internally by Dgraph. In this case, it's provided as an input because federated mutations aren't currently supported. The user should provide the same `id` value that is present in the GraphQL service where the type  `Astronaut` is defined.
```

> For example, let's take that the type `Astronaut` is defined in some other service `AstronautService` as -

Reviewer: ```suggestion
For example, let's assume that the type `Astronaut` is defined in some other service, `AstronautService`, as follows:
```

```graphql
type Astronaut @key(fields: "id") {
    id: ID! 
    name: String!
}
```

> When adding an object of type `Astronaut`, first it should be added into `AstronautService` service and then the `addAstronaut` mutation should be called with value of `id` provided as an argument which must be equal to the value in `AstronautService` service.

Reviewer: ```suggestion
When adding an object of type `Astronaut`, you should first add it to the `AstronautService` service. Then, you can call the `addAstronaut` mutation with the value of `id` provided as an argument that must be equal to the value in `AstronautService` service.
```
